### ISSUE TITLE
socket() for server.js

### ISSUE THREAD

**Issue by franciscop** [franciscop/server#63]
Checklist (progress on branch `plugins`):

- [x] Reworked request workflow to use `SOCKET` method
- [ ] Add options for `io()`
- [x] Handles no path as `*`
- [x] Multiple middleware
- [x] Defaults to emiting with the same name if you return anything

Thanks @paulrobello for showing me how you are using websockets and the struggle right now with server.js

The socket implementation is quite poor right now as its *Experimental* status should suggest, [but specially bad is the router](https://github.com/franciscop/server/blob/master/plugins/socket/index.js#L10):

```js
router: (path, middle) => {
  listeners[path] = listeners[path] || [];
  listeners[path].push(middle);
},
```

## Several middleware

This does not work right now:

```js
socket('a',
  () => { console.log('a'); },
  () => { console.log('b');
})
```

In fact, not even this will work [(fix)](https://stackoverflow.com/q/10405070):

```js
socket(() => { console.log('a'); });
```

This should be easy to fix [reusing other routers helper](https://github.com/franciscop/server/blob/master/router/parse.js) (has to be generalized first).

## Propagation

It doesn't behave like the other routes and errors don't come down. So this will not be able to handle the error:

```js
server({ ... },
  socket(ctx => { throw new Error('Websockets should be improved'); }),
  error(ctx => { console.log(ctx.error); })  // This does not work
);
```

The main issue is that a websocket request does not go through the whole middleware list. We could handle this several ways, but one of the ways I'm thinking about is to make a call into a normal request and route it through the whole stack with a `method = 'SOCKET'`. The compatibility layer `modern()` would skip these calls, so it would make the stack easier to work with:

```js
server(
  ctx => {
    // It could be either of these:
    // "GET undefined"
    // "POST { username: "franciscop" }"
    // "SOCKET { username: "franciscop" }"
    console.log(ctx.method, ctx.data);
  },
  get(),
  socket()
);
```

## Error propagation

Even fixing the two previous points, there is no error handling. This will fail catastrophically, potentially (not tested) stopping the server, which is something that should never happen with server.js:

```js
server({ ... },
  socket('a', () => {
    throw new Error('I break everything!');
  })
);
```

These are the main three issues right now. Basically, we need to make `socket()` into a first-class router citizen. If you are using `socket()` right now, chances are you will not need to change it, but in some cases you will need to change some things to accommodate this as explained.

## Breaking Changes

If you are already using `socket()`, these are the two dangerous areas:

**First**, and the most likely to affect you, is that now all plain middleware that is created specifically for server() will be called on a socket.emit event:

```js
server(ctx => {
  // Now this can also be 'SOCKET'! Beware
  // ctx. should be as similar as possible
  console.log(ctx.method);
});
```

This will not affect any express middleware since `modern()` will make sure not to call those for `SOCKET`.

> TODO: decide how the ctx and all the reply types affect middlewares when on the context of websockets. For instance, the replies `cookies()` and `jsonp()` don't even make sense for a `socket()` event.

**Second**, make sure to always use `socket()` within `server()` as in the documentation. This *does* work right now but it won't in the future:

```js
const server = require('server');
const { socket } = require('server/router');

// DO NOT DO THIS:
socket('hello', () => {
  console.log('Hello!');
});

server();
```

The reason is that `socket()` will become a normal router. Every example in the documentation made sure from the beginning to always use socket() within the other routes as I knew it would probably change.

> Note: this is the plan for sockets, however implementation limits might change things and I would update this and the official documentation accordingly.

**Comment by brumm** [comment]
Hey there,

I noticed this issue when I went spelunking for a little issue I'm having right now while attempting to use the (I know, unreleased) websocket plugin right now.

In my case, our server is unable to mount on '/' as per policy. All serverjs router handlers are used like so `get('/backend/foo', fn)`.

However, I can't do that for socketIO because the implementation is hidden in a plugin, and [no further options are passed to socketIO](https://github.com/franciscop/server/blob/master/plugins/socket/index.js#L15). In my specific case, I'd like to be able to use the [`path` option](https://github.com/franciscop/server/blob/master/plugins/socket/index.js#L15)

Stepping back from this specific problem, I guess I'd like to have some kind of `basePath` option available for serverjs, which would prepend a user-defined path for all routed paths, but that's for another issue.

**Comment by franciscop** [comment]
Very good point @brumm, this can and should be added as another option like any other plugin:

```js
server({
  socket: { ... }
});
```

Internally that object will then be passed to the second parameter of `io()`.

About the basePath, there is no option defined now nor will likely be. However I am considering this, might this work?

```js
server(mount('/backend', routes));
```

This would allow a certain amount of routes (or all in your situation) to be mounted on `/backend`. I do not like it, since I prefer that the routes are explicit and clear and I know it will be abused, but I can see the value for some specific cases.

**Comment by brumm** [comment]
Awesome! I'd love to be able to pass options to socketIO.

The basePath thing would really just be for convenience, and I agree that explicit is usually nicer. `mount()` would be a nice shortcut, but please don't feel like you should add more code just for my single use case.

**Comment by franciscop** [comment]
This should be available in the branch `plugins`. Unfortunately I want to wait until bundling it with the `1.1` version, so it is not ready yet. I have published it under the `1.1.0-alpha.1` version if you want to play around with the options for socket:

```bash
// .env
SOCKET_PATH=...
```

```js
server({ socket: { path: ... } }, ...);
```

If you find any bug please let me know! These options [should be the same as in the official socket.io documentation](https://socket.io/docs/server-api#new-server-httpserver-options).

**Comment by brumm** [comment]
Fantastic, thank you a lot! I'll give this a try.
